File: src/main/java/com/freya02/botcommands/internal/application/ApplicationCommandsUpdater.java
package com.freya02.botcommands.internal.application;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.application.CommandPath;
import com.freya02.botcommands.api.application.CommandScope;
import com.freya02.botcommands.api.builder.DebugBuilder;
import com.freya02.botcommands.internal.BContextImpl;
import com.freya02.botcommands.internal.application.context.message.MessageCommandInfo;
import com.freya02.botcommands.internal.application.context.user.UserCommandInfo;
import com.freya02.botcommands.internal.application.localization.BCLocalizationFunction;
import com.freya02.botcommands.internal.application.slash.SlashCommandInfo;
import com.freya02.botcommands.internal.application.slash.SlashUtils2;
import com.freya02.botcommands.internal.utils.Utils;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.interactions.commands.Command;
import net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions;
import net.dv8tion.jda.api.interactions.commands.build.*;
import net.dv8tion.jda.api.interactions.commands.localization.LocalizationFunction;
import net.dv8tion.jda.api.requests.restaction.CommandListUpdateAction;
import net.dv8tion.jda.internal.utils.Checks;
import org.jetbrains.annotations.Blocking;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.*;
import java.util.function.Function;

import static com.freya02.botcommands.internal.application.slash.SlashUtils.appendCommands;

@Deprecated
public class ApplicationCommandsUpdater {
	private static final Logger LOGGER = Logging.getLogger();

	private final BContextImpl context;
	@Nullable private final Guild guild;
	private final boolean onlineCheck;

	private final Path commandsCachePath;

	private final ApplicationCommandDataMap map = new ApplicationCommandDataMap();
	private final Map<String, SubcommandGroupData> subcommandGroupDataMap = new HashMap<>();
	private final List<ApplicationCommandInfo> guildApplicationCommands;

	private final List<Command> commands = new ArrayList<>();

	private final Collection<CommandData> allCommandData;

	private ApplicationCommandsUpdater(@NotNull BContextImpl context, @Nullable Guild guild, boolean onlineCheck) throws IOException {
		this.context = context;
		this.guild = guild;
		this.onlineCheck = onlineCheck;

		this.commandsCachePath = guild == null
				? context.getApplicationCommandsCache().getGlobalCommandsPath()
				: context.getApplicationCommandsCache().getGuildCommandsPath(guild);

		Files.createDirectories(commandsCachePath.getParent());

		final CommandIdProcessor commandIdProcessor = guild == null ? null : new CommandIdProcessor(context);
		this.guildApplicationCommands = this.context.getApplicationCommandsContext()
				.getApplicationCommandInfoMap()
				.filterByGuild(this.context, this.guild, commandIdProcessor);

		computeCommands();

		this.allCommandData = map.getAllCommandData();

		//Apply localization
		final LocalizationFunction localizationFunction = new BCLocalizationFunction(context);
		for (CommandData commandData : allCommandData) {
			commandData.setLocalizationFunction(localizationFunction);
		}
	}

	public static ApplicationCommandsUpdater ofGlobal(@NotNull BContextImpl context, boolean onlineCheck) throws IOException {
		return new ApplicationCommandsUpdater(context, null, onlineCheck);
	}

	public static ApplicationCommandsUpdater ofGuild(@NotNull BContextImpl context, @NotNull Guild guild, boolean onlineCheck) throws IOException {
		return new ApplicationCommandsUpdater(context, guild, onlineCheck);
	}

	public List<ApplicationCommandInfo> getGuildApplicationCommands() {
		return guildApplicationCommands;
	}

	@Nullable
	public Guild getGuild() {
		return guild;
	}

	@Blocking
	public boolean shouldUpdateCommands() throws IOException {
		final byte[] oldBytes;

		if (onlineCheck) {
			commands.clear();
			commands.addAll((guild == null ? context.getJDA().retrieveCommands(true) : guild.retrieveCommands(true)).complete());
			final List<CommandData> discordCommandsData = commands.stream().map(CommandData::fromCommand).toList();

			oldBytes = ApplicationCommandsCache.getCommandsBytes(discordCommandsData);
		} else {
			if (Files.notExists(commandsCachePath)) {
				LOGGER.trace("Updating commands because cache file does not exists");

				return true;
			}

			oldBytes = Files.readAllBytes(commandsCachePath);
		}

		final byte[] newBytes = ApplicationCommandsCache.getCommandsBytes(allCommandData);

		final boolean needUpdate = !ApplicationCommandsCache.isJsonContentSame(oldBytes, newBytes);

		if (needUpdate) {
			LOGGER.trace("Updating commands because content is not equal");

			if (DebugBuilder.isLogApplicationDiffsEnabled()) {
				LOGGER.trace("Old commands bytes: {}", new String(oldBytes));
				LOGGER.trace("New commands bytes: {}", new String(newBytes));
			}
		}

		return needUpdate;
	}

	@Blocking
	public void updateCommands() {
		final CommandListUpdateAction updateAction = guild != null ? guild.updateCommands() : context.getJDA().updateCommands();

		final List<Command> commands = updateAction
				.addCommands(allCommandData)
				.complete();

		if (guild != null) {
			thenAcceptGuild(commands, guild);
		} else {
			thenAcceptGlobal(commands);
		}
	}

	private void computeCommands() {
		computeSlashCommands(guildApplicationCommands);

		computeContextCommands(guildApplicationCommands, UserCommandInfo.class, Command.Type.USER);

		computeContextCommands(guildApplicationCommands, MessageCommandInfo.class, Command.Type.MESSAGE);
	}

	private void computeSlashCommands(List<ApplicationCommandInfo> guildApplicationCommands) {
		guildApplicationCommands.stream()
				.filter(a -> a instanceof SlashCommandInfo)
				.map(a -> (SlashCommandInfo) a)
				.forEachOrdered(info -> {
					final CommandPath commandPath = info.getPath();
					final String description = info.getDescription();

					try {
						final List<OptionData> methodOptions = SlashUtils2.getMethodOptions(info, context, guild);

						if (commandPath.getNameCount() == 1) {
							//Standard command
							final SlashCommandData rightCommand = Commands.slash(commandPath.getName(), description);
							map.put(Command.Type.SLASH, commandPath, rightCommand);

							rightCommand.addOptions(methodOptions);
							configureTopLevel(info, rightCommand);
						} else if (commandPath.getNameCount() == 2) {
							Checks.notNull(commandPath.getSubname(), "Subcommand name");

							final SlashCommandData commandData = (SlashCommandData) map.computeIfAbsent(Command.Type.SLASH, commandPath, x -> {
								final SlashCommandData tmpData = Commands.slash(commandPath.getName(), "No description (base name)");
								configureTopLevel(info, tmpData);

								return tmpData;
							});

							//Subcommand of a command
							final SubcommandData subcommandData = new SubcommandData(commandPath.getSubname(), description);
							subcommandData.addOptions(methodOptions);

							commandData.addSubcommands(subcommandData);
						} else if (commandPath.getNameCount() == 3) {
							Checks.notNull(commandPath.getGroup(), "Command group name");
							Checks.notNull(commandPath.getSubname(), "Subcommand name");

							final SubcommandGroupData groupData = getSubcommandGroup(Command.Type.SLASH, commandPath, x -> {
								final SlashCommandData commandData = Commands.slash(commandPath.getName(), "No description (base name)");
								configureTopLevel(info, commandData);

								return commandData;
							});

							final SubcommandData subcommandData = new SubcommandData(commandPath.getSubname(), description);
							subcommandData.addOptions(methodOptions);

							groupData.addSubcommands(subcommandData);
						} else {
							throw new IllegalStateException("A slash command with more than 4 path components got registered");
						}
					} catch (Exception e) {
						throw new RuntimeException("An exception occurred while processing command '" + commandPath + "' at " + Utils.formatMethodShort(info.getMethod()), e);
					}
				});
	}

	@SuppressWarnings("unchecked")
	private <T extends ApplicationCommandInfo> void computeContextCommands(List<ApplicationCommandInfo> guildApplicationCommands, Class<T> targetClazz, Command.Type type) {
		guildApplicationCommands.stream()
				.filter(a -> targetClazz.isAssignableFrom(a.getClass()))
				.map(a -> (T) a)
				.forEachOrdered(info -> {
					final CommandPath commandPath = info.getPath();

					try {
						if (commandPath.getNameCount() == 1) {
							//Standard command
							final CommandData rightCommand = Commands.context(type, commandPath.getName());
							map.put(type, commandPath, rightCommand);

							configureTopLevel(info, rightCommand);
						} else {
							throw new IllegalStateException("A " + type.name() + " command with more than 1 path component got registered");
						}
					} catch (Exception e) {
						throw new RuntimeException("An exception occurred while processing a " + type.name() + " command " + commandPath, e);
					}
				});
	}

	private void configureTopLevel(ApplicationCommandInfo info, CommandData rightCommand) {
		if (info.isDefaultLocked()) {
			rightCommand.setDefaultPermissions(DefaultMemberPermissions.DISABLED);
		} else if (!info.getUserPermissions().isEmpty()) {
			rightCommand.setDefaultPermissions(DefaultMemberPermissions.enabledFor(info.getUserPermissions()));
		}

		if (info.getScope() == CommandScope.GLOBAL_NO_DM) {
			rightCommand.setGuildOnly(true);
		}
	}

	private void thenAcceptGuild(List<Command> commands, @NotNull Guild guild) {
		for (Command command : commands) {
			context.getRegistrationListeners().forEach(l -> l.onGuildSlashCommandRegistered(this.guild, command));
		}

		this.commands.clear();
		this.commands.addAll(commands);

		try {
			Files.write(commandsCachePath, ApplicationCommandsCache.getCommandsBytes(allCommandData), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
		} catch (IOException e) {
			LOGGER.error("An exception occurred while temporarily saving guild ({} ({})) commands in '{}'", guild.getName(), guild.getId(), commandsCachePath.toAbsolutePath(), e);
		}

		if (!LOGGER.isTraceEnabled()) return;

		final StringBuilder sb = new StringBuilder("Updated " + commands.size() + " / " + allCommandData.size() + " (" + context.getApplicationCommandsView().size() + ") commands for ");
		sb.append(guild.getName()).append(" :\n");
		appendCommands(commands, sb);

		LOGGER.trace(sb.toString().trim());
	}

	private void thenAcceptGlobal(List<Command> commands) {
		for (Command command : commands) {
			context.getRegistrationListeners().forEach(l -> l.onGlobalSlashCommandRegistered(command));
		}

		try {
			Files.write(commandsCachePath, ApplicationCommandsCache.getCommandsBytes(allCommandData), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
		} catch (IOException e) {
			LOGGER.error("An exception occurred while temporarily saving {} commands in '{}'", guild == null ? "global" : String.format("guild '%s' (%s)", guild.getName(), guild.getId()), commandsCachePath.toAbsolutePath(), e);
		}

		if (!LOGGER.isTraceEnabled()) return;

		final StringBuilder sb = new StringBuilder("Updated global commands:\n");
		appendCommands(commands, sb);

		LOGGER.trace(sb.toString().trim());
	}

	//I am aware that the type is always Command.Type#SLASH, still use a parameter to mimic how ApplicationCommandMap functions and for future proof uses
	@SuppressWarnings("SameParameterValue")
	@NotNull
	private SubcommandGroupData getSubcommandGroup(Command.Type type, CommandPath path, Function<String, CommandData> baseCommandSupplier) {
		if (path.getGroup() == null)
			throw new IllegalArgumentException("Group component of command path is null at '" + path + "'");

		final SlashCommandData data = (SlashCommandData) map.computeIfAbsent(type, path, baseCommandSupplier);

		final CommandPath parent = path.getParent();
		if (parent == null) throw new IllegalStateException("A command path with less than 3 components was passed to #getSubcommandGroup");

		return subcommandGroupDataMap.computeIfAbsent(parent.getFullPath(), s -> {
			final SubcommandGroupData groupData = new SubcommandGroupData(path.getGroup(), "No description (group)");

			data.addSubcommandGroups(groupData);

			return groupData;
		});
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/commands/application/ApplicationCommandsUpdater.kt
package com.freya02.botcommands.internal.commands.application

import com.freya02.botcommands.api.commands.application.AbstractApplicationCommandManager
import com.freya02.botcommands.api.commands.application.CommandScope
import com.freya02.botcommands.api.commands.application.GlobalApplicationCommandManager
import com.freya02.botcommands.api.commands.application.GuildApplicationCommandManager
import com.freya02.botcommands.api.core.service.getService
import com.freya02.botcommands.api.core.utils.overwriteBytes
import com.freya02.botcommands.internal.BContextImpl
import com.freya02.botcommands.internal.commands.application.ApplicationCommandsCache.Companion.toJsonBytes
import com.freya02.botcommands.internal.commands.application.context.message.MessageCommandInfo
import com.freya02.botcommands.internal.commands.application.context.user.UserCommandInfo
import com.freya02.botcommands.internal.commands.application.localization.BCLocalizationFunction
import com.freya02.botcommands.internal.commands.application.mixins.ITopLevelApplicationCommandInfo
import com.freya02.botcommands.internal.commands.application.slash.SlashSubcommandGroupInfo
import com.freya02.botcommands.internal.commands.application.slash.SlashSubcommandInfo
import com.freya02.botcommands.internal.commands.application.slash.SlashUtils.getDiscordOptions
import com.freya02.botcommands.internal.commands.application.slash.TopLevelSlashCommandInfo
import com.freya02.botcommands.internal.commands.mixins.INamedCommand
import com.freya02.botcommands.internal.utils.asScopeString
import com.freya02.botcommands.internal.utils.rethrowUser
import dev.minn.jda.ktx.coroutines.await
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import mu.KotlinLogging
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.interactions.commands.Command
import net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions
import net.dv8tion.jda.api.interactions.commands.build.CommandData
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.dv8tion.jda.api.interactions.commands.build.SubcommandData
import net.dv8tion.jda.api.interactions.commands.build.SubcommandGroupData
import net.dv8tion.jda.api.interactions.commands.localization.LocalizationFunction
import java.nio.file.Files

internal class ApplicationCommandsUpdater private constructor(
    private val context: BContextImpl,
    private val guild: Guild?,
    manager: AbstractApplicationCommandManager
) {
    private val logger = KotlinLogging.logger {  }

    private val commandsCache = context.getService<ApplicationCommandsCache>()
    private val onlineCheck = context.applicationConfig.onlineAppCommandCheckEnabled

    private val commandsCachePath = when (guild) {
        null -> commandsCache.globalCommandsPath
        else -> commandsCache.getGuildCommandsPath(guild)
    }

    val applicationCommands: Collection<ApplicationCommandInfo>
    private val allCommandData: Collection<CommandData>

    init {
        Files.createDirectories(commandsCachePath.parent)

        applicationCommands = manager.applicationCommands.values
        allCommandData = computeCommands().allCommandData

        //Apply localization
        val localizationFunction: LocalizationFunction = BCLocalizationFunction(context)
        for (commandData in allCommandData) {
            commandData.setLocalizationFunction(localizationFunction)
        }
    }

    suspend fun shouldUpdateCommands(): Boolean {
        val oldBytes = when {
            onlineCheck -> {
                (guild?.retrieveCommands(true) ?: context.jda.retrieveCommands(true))
                    .await()
                    .map { CommandData.fromCommand(it) }.toJsonBytes()
            }

            else -> {
                if (Files.notExists(commandsCachePath)) {
                    logger.trace("Updating commands because cache file does not exists")
                    return true
                }

                withContext(Dispatchers.IO) {
                    Files.readAllBytes(commandsCachePath)
                }
            }
        }

        val newBytes = allCommandData.toJsonBytes()
        return (!ApplicationCommandsCache.isJsonContentSame(context, oldBytes, newBytes)).also { needUpdate ->
            if (needUpdate) {
                logger.trace("Updating commands because content is not equal")

                if (context.debugConfig.enableApplicationDiffsLogs) {
                    logger.trace { "Old commands bytes: ${oldBytes.decodeToString()}" }
                    logger.trace { "New commands bytes: ${newBytes.decodeToString()}" }
                }
            }
        }
    }

    suspend fun updateCommands() {
        val updateAction = guild?.updateCommands() ?: context.jda.updateCommands()
        val commands = updateAction
            .addCommands(allCommandData)
            .await()

        saveCommandData(guild)
        printPushedCommandData(commands, guild)
    }

    private fun computeCommands() = ApplicationCommandDataMap().also { map ->
        computeSlashCommands(applicationCommands, map)
        computeContextCommands(applicationCommands, map, UserCommandInfo::class.java, Command.Type.USER)
        computeContextCommands(applicationCommands, map, MessageCommandInfo::class.java, Command.Type.MESSAGE)
    }

    private fun computeSlashCommands(guildApplicationCommands: Collection<ApplicationCommandInfo>, map: ApplicationCommandDataMap) {
        guildApplicationCommands
            .filterIsInstance<TopLevelSlashCommandInfo>()
            .filterCommands()
            .forEach { info: TopLevelSlashCommandInfo ->
                try {
                    val isTopLevel = info.isTopLevelCommandOnly()
                    val topLevelData = Commands.slash(info.name, info.description).also { commandData ->
                        if (isTopLevel) {
                            commandData.addOptions(info.getDiscordOptions(guild))
                        }

                        commandData.configureTopLevel(info)
                    }

                    topLevelData.addSubcommandGroups(info.subcommandGroups.values.filterCommands().mapToSubcommandGroupData())
                    topLevelData.addSubcommands(info.subcommands.values.filterCommands().mapToSubcommandData())

                    map[Command.Type.SLASH, info.name] = topLevelData
                } catch (e: Exception) { //TODO use some sort of exception context for command paths
                    rethrowUser(info.function, "An exception occurred while processing command '${info.name}'", e)
                }
            }
    }

    private fun Collection<SlashSubcommandGroupInfo>.mapToSubcommandGroupData() =
        this.map { subcommandGroupInfo ->
            SubcommandGroupData(subcommandGroupInfo.name, subcommandGroupInfo.description).also {
                it.addSubcommands(subcommandGroupInfo.subcommands.values.mapToSubcommandData())
            }
        }

    private fun Collection<SlashSubcommandInfo>.mapToSubcommandData() =
        this.map { subcommandInfo ->
            SubcommandData(subcommandInfo.name, subcommandInfo.description)
                .addOptions(subcommandInfo.getDiscordOptions(guild))
        }

    private fun <T> computeContextCommands(
        guildApplicationCommands: Collection<ApplicationCommandInfo>,
        map: ApplicationCommandDataMap,
        targetClazz: Class<T>,
        type: Command.Type
    ) where T : ITopLevelApplicationCommandInfo,
            T : ApplicationCommandInfo {
        guildApplicationCommands
            .filterIsInstance(targetClazz)
            .filterCommands()
            .forEach { info: T ->
                try {
                    //Standard command
                    map[type, info.name] = Commands.context(type, info.name).configureTopLevel(info)
                } catch (e: Exception) {
                    rethrowUser(info.function, "An exception occurred while processing a ${type.name} command ${info.name}", e)
                }
            }
    }

    private fun <T : INamedCommand> Collection<T>.filterCommands() = filter { info ->
        context.settingsProvider?.let { settings ->
            guild?.let { guild ->
                return@filter settings.getGuildCommands(guild).filter.test(info.path)
            }
        }

        return@filter true
    }

    private fun <T> CommandData.configureTopLevel(info: T): CommandData
            where T : ITopLevelApplicationCommandInfo,
                  T : ApplicationCommandInfo = apply {
        if (info.nsfw) isNSFW = true
        if (info.scope == CommandScope.GLOBAL_NO_DM) isGuildOnly = true
        if (info.isDefaultLocked) {
            defaultPermissions = DefaultMemberPermissions.DISABLED
        } else if (info.userPermissions.isNotEmpty()) {
            defaultPermissions = DefaultMemberPermissions.enabledFor(info.userPermissions)
        }
    }

    private fun printPushedCommandData(commands: List<Command>, guild: Guild?) {
        if (!logger.isTraceEnabled) return

        logger.trace {
            val commandNumber = commands.size
            val sentCommandNumber = allCommandData.size
            val scope = guild.asScopeString()
            "Updated $commandNumber / $sentCommandNumber commands for $scope"
        }
    }

    private fun saveCommandData(guild: Guild?) {
        try {
            commandsCachePath.overwriteBytes(allCommandData.toJsonBytes())
        } catch (e: Exception) {
            logger.error(e) {
                "An exception occurred while temporarily saving ${guild.asScopeString()} commands in '${commandsCachePath.toAbsolutePath()}'"
            }
        }
    }

    companion object {
        fun ofGlobal(context: BContextImpl, manager: GlobalApplicationCommandManager): ApplicationCommandsUpdater {
            return ApplicationCommandsUpdater(context, null, manager)
        }

        fun ofGuild(context: BContextImpl, guild: Guild, manager: GuildApplicationCommandManager): ApplicationCommandsUpdater {
            return ApplicationCommandsUpdater(context, guild, manager)
        }
    }
}